src/Storages/MergeTree/checkDataPart.cpp

```diff
@@ -201,19 +201,21 @@ static IMergeTreeDataPart::Checksums checkDataPart(
             continue;
 
         auto checksum_it = checksums_data.files.find(file_name);
-
         /// Skip files that we already calculated. Also skip metadata files that are not checksummed.
         if (checksum_it == checksums_data.files.end() && !files_without_checksums.contains(file_name))
         {
             auto txt_checksum_it = checksums_txt_files.find(file_name);
-            if (txt_checksum_it == checksums_txt_files.end() || txt_checksum_it->second.uncompressed_size == 0)
+            if ((txt_checksum_it != checksums_txt_files.end() && txt_checksum_it->second.uncompressed_size != 0)
+                || file_name.contains(".cmrk")
+                || file_name.ends_with(".cidx"))
```

Comment:
Does it mean that we forgot to set `uncompressed_size` for compressed cmrk/cidx files when calculating checksums?

Reply:
When we calculate and write checksums, we write correctly. 
for cmrk file
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeDataPartWriterCompact.cpp#L378-L383
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeDataPartWriterOnDisk.cpp#L91-L96

for cidx
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeDataPartWriterOnDisk.cpp#L363-L368

But during `checkDataPart` we didn't consider them as compressed. So we would have `uncompressed_size` written, but not calculated here. Thus it would fail in `checkEqual`
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeDataPartChecksum.cpp#L33-L34

Reply:
ya, I get your point. It should have passed '`txt_checksum_it->second.uncompressed_size != 0`'. I think we should check `is_compressed` & not size.

Reply:
Aah, I got it, so the problem is that the size may actually be 0